content/ui-patterns/forms.mdx

```diff
@@ -0,0 +1,295 @@
+    <Caption>Don’t show a repetitive validation message</Caption>
+  </Dont>
+</DoDontContainer>
+
+For more information about form validation, see the [validation guidelines](#invalid-input).
+
+## Input methods
+
+### Open-ended text
+
+![Text input, open-ended autocomplete text input, plain text input, textarea](https://user-images.githubusercontent.com/2313998/164782311-93a743fb-7757-4f29-bc74-522408a01b5c.png)
+
+Use an open-ended text field when the field does not have a list of possible values. If it's possible to suggest values, use an [Autocomplete](#autocomplete) that allows users to pick a value or enter their own.
+
+### A set of selectable options
+
+![Action menu, autocomplete input with limited options, checkbox group, radio group, select](https://user-images.githubusercontent.com/2313998/164782313-851a5a90-8c50-409f-b03d-4200928e0a67.png)
+
+Show a set of selectable options when there is a finite number of possible values.
+
+## Structure
+
+Forms should have a structure that makes it easy for users to scan. Form controls that are stacked vertically and left-aligned are easier for sighted users to scan visually.
+
+<DoDontContainer>
+  <Do>
+    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065202-5c3aba0b-5cef-4ef4-9339-7192a3ebf23a.png" />
+    <Caption>Default to vertically stacked form controls</Caption>
+  </Do>
+  <Dont>
+    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065205-4000430d-2216-4954-a367-86fbd0254cc3.png" />
+    <Caption>Don’t lay out forms into columns just to reduce the vertical space used by the form</Caption>
+  </Dont>
+</DoDontContainer>
+
+### Order of form controls
+
+Form fields should be in a predictable order that flows intuitively.
+
+To achieve an intuitive flow:
+- Order fields by their relative level of importance
+- Keep related fields near each other
+- When possible, keep inputs that require keyboard input near each other so that users who use a mouse don't have to switch back and forth between clicking and typing
+
+### Grouping form controls
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
+        width="456"
+        alt="visually grouped of related inputs"
+        src="https://user-images.githubusercontent.com/2313998/164065207-16e51727-55b2-48bc-930c-ad9f0890ae10.png"
+    />
+    <Box as="p" mt="0">When there is a collection of closely related fields, they should be labeled and visually grouped together. Effective techniques to visually group a collection of form controls include proximity and whitespace.</Box>
+</Box>
+
+### Progressively disclosed form controls
+
+<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
+    <img
```

Comment:
> <img src="https://user-images.githubusercontent.com/293280/166303136-cc5d6cdb-494d-4836-8137-7c0930d36e71.png" alt="" width="400" />

Is this a realistic width we should recommend for controls? Everything looks so squished horizontally. Also, the spacing between expiration and CVV feels incorrect. 

Additionally, shouldn't these controls be grouped, as described in the guidelines right above this section?

Reply:
> Everything looks so squished horizontally.

I think everything looks so squished because of the text in the diagram. 

---

> Also, the spacing between expiration and CVV feels incorrect.

It definitely is. My mistake.

---

> Additionally, shouldn't these controls be grouped, as described in the guidelines right above this section?

Yup, another mistake on my part.